Implement FixedFootDetector class

GiulioRomualdi opened the pull request:
This PR implements the same logic used in the unicycle library to find the fixed foot given a list of planned contact.

Before reviewing the pr I have to:
- [x] write the test
- [x] Improve the documentation 


This code is meant to be used with the leggedodomerty class to get the robot base from a list of planned contact. 

cc @S-Dafarra and @prashanthr05

S-Dafarra commented:
Initial picky comment. The name ``BaseDetector`` seems a bit off. It does not detect the base, as that is usually a choice. ``FixedFootDetector``?

GiulioRomualdi commented:
I also don't like the name. I don't know if `FootFixedDetector` is the right choice since also the Smith trigger detector aims to find the stance foot. 

Probably @prashanthr05  has some idea

prashanthr05 commented:
> I also don't like the name. I don't know if `FootFixedDetector` is the right choice since also the Smith trigger detector aims to find the stance foot.
> 
> Probably @prashanthr05 has some idea

`SchmittTriggerDetector` just tries to detect if any given link is in contact or not. It does not make decision on the stance foot.

prashanthr05 commented:
Maybe we could eventually move these methods ithat currently exist in LeggedOdometry to determine the latest switched contact and the last active contact from a list of contacts using a switching time parameter. also into this class.

https://github.com/dic-iit/bipedal-locomotion-framework/blob/4ad31d8d62ffa0e3da5b7fbef77a945124c7bb4e/src/Estimators/src/LeggedOdometry.cpp#L336-L409

GiulioRomualdi commented:
Documentation and test written. I also renamed the class as suggested by @S-Dafarra.
The PR is ready to be reviewed

S-Dafarra commented on src/Contacts/src/FixedFootDetector.cpp:
```diff
@@ -0,0 +1,213 @@
+    if (phase == m_contactPhaselist.cend())
+        phase = m_contactPhaselist.lastPhase();
+
+    // if the robot is in single support only once contact is enabled
```
```suggestion
    // if the robot is in single support only one contact is enabled
```

S-Dafarra commented on src/Contacts/src/FixedFootDetector.cpp:
```diff
@@ -0,0 +1,213 @@
+        }
+    } else if (phase->activeContacts.size() == 2)
+    {
+        // Notice that here we do not analyze the case in which the phase is differnt from
```
```suggestion
        // Notice that here we do not analyze the case in which the phase is different from
```

prashanthr05 commented on src/Contacts/include/BipedalLocomotion/ContactDetectors/ContactDetector.h:
```diff
@@ -8,105 +8,102 @@
-        Initialized,    /**< The estimator is initialized and ready to be used */
-        Running         /**< The estimator is running */
+        NotInitialized, /**< The estimator is not initialized yet call
+                           FloatingBaseEstimator::initialze method to initialize it*/
```
Leftover from previous PRs 
```suggestion
                           ContactDetector::initialize method to initialize it*/
```

prashanthr05 commented on src/Contacts/include/BipedalLocomotion/ContactDetectors/FixedFootDetector.h:
```diff
@@ -0,0 +1,88 @@
+ * R            |+++++++++++|---|+++++++++++|---|+++++++++++|---|+++++++++++|---|+++|
+ * stance foot  |LLL|RRR|RRR|LLL|LLL|RRR|RRR|LLL|LLL|RRR|RRR|LLL|LLL|RRR|RRR|LLL|LLL|
+ * @endcode
+ * @warning The FixedFootDetector can handle only bipedal locomotion. Furthermore only one foot is
+ * considered as fixed foot
```
Let's document this assumption more in detail maybe?

Saying that, the implementation assumes that the contact phase list only expects two contacts data and that at least one of the contact must be active at all times?

prashanthr05 commented on src/Contacts/src/FixedFootDetector.cpp:
```diff
@@ -0,0 +1,213 @@
+    {
+        // Notice that here we do not analyze the case in which the phase is differnt from
+        // firstPhase. Indeed in this case we should not update the contact states dictionary. (The
+        // contet of the dictionary is the one of the previous single support phase)
```
```suggestion
        // content of the dictionary is the one of the previous single support phase)
```

prashanthr05 commented on src/Contacts/src/FixedFootDetector.cpp:
```diff
@@ -0,0 +1,213 @@
+        // contet of the dictionary is the one of the previous single support phase)
+
+        // if the current phase is the first and there are at least 2 phases the active contact will
+        // be the one that in the next phase is active
```
```suggestion
        // be the one that is going to be active in the next phase
```

prashanthr05 commented on src/Contacts/tests/ContactDetectors/FixedFootDetectorUnitTest.cpp:
```diff
@@ -0,0 +1,168 @@
+
+struct FixedFootState
+{
+    EstimatedContact leftFoot;
+    EstimatedContact rightFoot;
```
Should this be a `PlannedContact` instead of an `EstimatedContact`?

prashanthr05 replied:
Nevermind, this is on the output end of the planner and input end of the estimator... so, `EstimatedContact` is fitting.

prashanthr05 reviewed (commented):
Minor Comments.

GiulioRomualdi commented:
Once https://github.com/dic-iit/bipedal-locomotion-framework/pull/297 will be merged the tests shouldn't fail anymore

GiulioRomualdi commented:
Hi @prashanthr05 and @S-Dafarra, in https://github.com/dic-iit/bipedal-locomotion-framework/pull/284/commits/113cfb9576f318882e339823775d5b18ba91cb51 I updated the logic used to generate the fixed foot. 

I also updated the documentation. Could you please check the PR?
